### PR TITLE
test: (kernel) submission validation

### PR DESCRIPTION
**Pull request by lordmallam**
### Validates a submission payload using the provided `mappingset_id`
Reachable at `POST /submissions/validate/`

### JIRA ticket: [AET-468](https://jira.ehealthafrica.org/browse/AET-468)
expected body:
            ```{
                'mappingset': 'uuid',
                'payload': {}
            }```

expected response:
        ```{
            'is_valid': 'true||false',
            'entities': [],
            aether_errors: []
        }```

### CODE REVIEW

File: aether-kernel/aether/kernel/api/views.py [hunk 1]
```diff
@@ -359,6 +360,72 @@ class SubmissionViewSet(MtViewSetMixin, ExporterViewSet):
...
+
+        mappingset_id = request.data.get('mappingset')
+        payload = request.data.get('payload')
+        if mappingset_id is None or payload is None:
+            return Response(
+                _('A mappingset id and payload must be provided'),
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        mappings = models.MappingSet.objects.get(pk=mappingset_id).mappings.all()
```
Comment (obdulia-losantos) [inline aether-kernel/aether/kernel/api/views.py]: do not used this method if multitenancy is enabled, it breaks security  :scream: 

```python
from django.shortcuts import get_object_or_404
from rest_framework.exceptions import PermissionDenied
from aether.django.sdk.multitenancy.utils import is_accessible_by_realm

mappingset = get_object_or_404(models.MappingSet.objects.all(), pk=mappingset_id)
if not is_accessible_by_realm(request, mappingset):
    raise PermissionDenied(_('Not accessible by this realm'))

mappings = mappingset.mappings.all()
```

**Review by obdulia-losantos — CHANGES_REQUESTED** [review]
I tiny security hole... :hole: :golfing_man:

File: aether-kernel/aether/kernel/api/views.py [hunk 2]
```diff
@@ -396,7 +397,12 @@ def validate(self, request, *args, **kwargs):
                 status=status.HTTP_400_BAD_REQUEST,
             )
 
-        mappings = models.MappingSet.objects.get(pk=mappingset_id).mappings.all()
+        mappingset = get_object_or_404(models.MappingSet.objects.all(), pk=mappingset_id)
+        x = is_accessible_by_realm(request, mappingset)
+        print('X', x)
+        if not x:
```
Comment (obdulia-losantos) [inline aether-kernel/aether/kernel/api/views.py]: :joy: 

Debug enabled :stuck_out_tongue_closed_eyes:

Comment (lordmallam) [reply aether-kernel/aether/kernel/api/views.py]: 🤣

File: aether-kernel/aether/kernel/api/views.py [hunk 3]
```diff
@@ -359,6 +361,75 @@ class SubmissionViewSet(MtViewSetMixin, ExporterViewSet):
...
+            schemas = {}
+            for sd in mapping.schemadecorators.all():
+                schemas[sd.name] = sd.schema.definition
+            try:
+                submission_data, entities = extract_create_entities(
+                    submission_payload=payload,
+                    mapping_definition=mapping.definition,
+                    schemas=schemas,
+                )
+                if ENTITY_EXTRACTION_ERRORS in submission_data and len(submission_data[ENTITY_EXTRACTION_ERRORS]):
```
Comment (obdulia-losantos) [inline aether-kernel/aether/kernel/api/views.py]: python shortcut :woman_shrugging: 

```python
if submission_data.get(ENTITY_EXTRACTION_ERRORS):
```

If ENTITY_EXTRACTION_ERRORS is not in submission_data returns `None` then `False`
If submission_data[ENTITY_EXTRACTION_ERRORS] is an empty list then `False`
else `True`

_what we always forgot about python, chapter 3_

File: aether-kernel/aether/kernel/api/views.py [hunk 4]
```diff
@@ -359,6 +361,75 @@ class SubmissionViewSet(MtViewSetMixin, ExporterViewSet):
...
+        if not is_accessible_by_realm(request, mappingset):
+            raise PermissionDenied(_('Not accessible by this realm'))
+        mappings = mappingset.mappings.all()
+        result = {
+            'is_valid': True,
+            'entities': [],
+            ENTITY_EXTRACTION_ERRORS: []
+        }
+        for mapping in mappings:
+            schemas = {}
```
Comment (obdulia-losantos) [inline aether-kernel/aether/kernel/api/views.py]: Another shortcut _chapter 4_

```python
schemas = {
    sd.name: sd.schema.definition
    for sd in mapping.schemadecorators.all()
}

```

**Review by obdulia-losantos — DISMISSED** [review]
Note: read _chapter 5_ and explain it to me :books:  :wink:

**Review by obdulia-losantos — APPROVED** [review]

**Pull request merged**